src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -246,6 +250,63 @@ void QueryToOperatorTransformer::Visit(parser::InsertStatement *op) {
     insert_expr->PushChild(output_expr_);
     output_expr_ = insert_expr;
   } else {
+    /*auto storage_manager = storage::StorageManager::GetInstance();
+    auto database =
+        storage_manager->GetDatabaseWithOid(target_table->GetDatabaseOid());
+    auto table = database->GetTableWithOid(target_table->GetTableOid());
+
+    auto *schema = table->GetSchema();
+    */
+
+    // column_objects represents the columns for the current table as defined in
+    // its schema
+    auto column_objects = target_table->GetColumnObjects();
+    // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+    if (op->columns.empty()) {
+      for (uint32_t tuple_idx = 0; tuple_idx < op->insert_values.size();
+           tuple_idx++) {
+        auto &values = (op->insert_values)[tuple_idx];
+        if (values.size() > column_objects.size()) {
+          throw CatalogException(
+              "ERROR:  INSERT has more expressions than target columns");
+        } else if (values.size() < column_objects.size()) {
+          throw CatalogException(
```

Comment:
Here we need to check if the remaining columns allow for null values.

Reply:
@schedutron Correct. You also need to check if the inserted type is "compatible" with the column type. Also, try to ensure you have the same error message as Postgres for these types of errors.

Reply:
@pmenon  I'm working on the null part, but I've added some compatibility checks in #1088. Do I need to add them here as well?